File: week 2/toposort/Toposort.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class Toposort {
    private static ArrayList<Integer> toposort(ArrayList<Integer>[] adj) {
        int used[] = new int[adj.length];
        ArrayList<Integer> order = new ArrayList<Integer>();
        for (int i = 0; i < adj.length; i++) {
			if (used[i] == 0) {
				dfs(adj, used, order, i);
			}
		}
        return order;
    }

    private static void dfs(ArrayList<Integer>[] adj, int[] used, ArrayList<Integer> order, int s) {
      used[s] = 1;
	  for (int i = 0; i < adj[s].size(); i++) {
		  if (used[adj[s].get(i)] == 0) {
			   dfs(adj, used, order, adj[s].get(i));
		  }
	  }
	  order.add(0, s);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
        }
        ArrayList<Integer> order = toposort(adj);
        for (int x : order) {
            System.out.print((x + 1) + " ");
        }
    }
}

File: week 2/toposort/toposort.py
#Uses python3

import sys

def dfs(adj, used, order, x):
    used[x] = 1
    for i in range(len(adj[x])):
        if not used[adj[x][i]]:
            dfs(adj, used, order, adj[x][i])
    order.append(x)


def toposort(adj):
    used = [0] * len(adj)
    order = []
    for i in range(len(adj)):
        if not used[i]:
            dfs(adj, used, order, i)
    order.reverse()
    return order

if __name__ == '__main__':
    n, m = map(int, raw_input().split())
    adj = [[] for _ in range(n)]
    for i in range(m):
        a, b = map(int, raw_input().split())
        # adjacency list		
        adj[a - 1].append(b - 1)
    order = toposort(adj)
    for x in order:
        print x + 1,

